libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,690 @@
+    const float roll = RC.get_roll_channel().norm_input_dz();
+    const float pitch = RC.get_pitch_channel().norm_input_dz();
+    const float yaw = RC.get_yaw_channel().norm_input_dz();
+
+    if (fabsf(roll) > 0 || fabsf(pitch) > 0 || fabsf(yaw) > 0) {
+        return true;
+    }
+    return false;
+}
+
+// Get the axis name we are working on, or DONE for all done 
+AP_Quicktune::AxisName AP_Quicktune::get_current_axis() const
+{
+    for (int8_t i = 0; i < int8_t(AxisName::DONE); i++) {
+        if (BIT_IS_SET(axes_enabled, i) == true && BIT_IS_SET(axes_done, i) == false) {
+            return AxisName(i);
+        }
+    }
+    return AxisName::DONE;
+}
+
+// get the AC_PID object for an axis
+AC_PID *AP_Quicktune::get_axis_pid(AP_Quicktune::AxisName axis) const
+{
+    auto &attitude_control = *AC_AttitudeControl::get_singleton();
+    switch(axis) {
+    case AxisName::RLL:
+        return &attitude_control.get_rate_roll_pid();
+    case AxisName::PIT:
+        return &attitude_control.get_rate_pitch_pid();
+    case AxisName::YAW:
+        return &attitude_control.get_rate_yaw_pid();
+    default:
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        break;
+    }
+    return nullptr;
+}
+
+// get slew rate parameter for an axis
+float AP_Quicktune::get_slew_rate(AP_Quicktune::AxisName axis) const
+{
+    auto *pid = get_axis_pid(axis);
+    if (pid == nullptr) {
+        return 0;
+    }
+    return pid->get_pid_info().slew_rate;
+}
+
+// Move to next stage of tune
+void AP_Quicktune::advance_stage(AP_Quicktune::AxisName axis)
+{
+    if (current_stage == Stage::D) {
+        current_stage = Stage::P;
+    } else {
+        BIT_SET(axes_done, uint8_t(axis));
+        GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Quicktune: %s done", get_axis_name(axis));
+        current_stage = Stage::D;
+    }
+}
+
+void AP_Quicktune::adjust_gain(AP_Quicktune::Param param, float value)
+{
+    need_restore = true;
+    BIT_SET(param_changed, uint8_t(param));
+    set_param_value(param, value);
+
+    if (get_stage(param) == Stage::P) {
+        // Also change I gain
+        const uint8_t P_TO_I_OFS = uint8_t(Param::RLL_I) - uint8_t(Param::RLL_P);
+        const uint8_t P_TO_FF_OFS = uint8_t(Param::RLL_FF) - uint8_t(Param::RLL_P);
+        const Param iname = Param(uint8_t(param)+P_TO_I_OFS);
+        const Param ffname = Param(uint8_t(param)+P_TO_FF_OFS);
+        float FF = get_param_value(ffname);
+        if (FF > 0) {
+            // If we have any FF on an axis then we don't couple I to P,
+            // usually we want I = FF for a one second time constant for trim
+            return;
+        }
+        BIT_SET(param_changed, uint8_t(iname));
+
+        // Work out ratio of P to I that we want
+        float pi_ratio = rp_pi_ratio;
+        if (get_axis(param) == AxisName::YAW) {
+            pi_ratio = y_pi_ratio;
+        }
+        if (pi_ratio >= 1) {
+            set_param_value(iname, value/pi_ratio);
```

Comment:
```suggestion
        if (pi_ratio <= 1) {
            set_param_value(iname, value * pi_ratio);
```
This is what I would have expected.
Or:
```
set_param_value(iname, value * MIN(pi_ratio, 1.0);
```

Reply:
it was the parameter range that was wrong, the code was right. Using a smaller I than P is allowed, but I greater than P will need manual tuning. Just too easy to get that wrong (I've seen what happens with I > P on a large quadplane)